go/atomic2/atomic128_amd64.s

```diff
@@ -0,0 +1,46 @@
+// Copyright 2023 The Vitess Authors.
+// Copyright (c) 2021, Carlo Alberto Ferraris
+// Copyright (c) 2017, Tom Thorogood
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Use of this source code is governed by a
+// Modified BSD License that can be found in
+// the LICENSE file.
+
+//+build !noasm,!appengine
+
+#include "textflag.h"
+
+TEXT ·compareAndSwapUint128_(SB), NOSPLIT, $0-41
+	MOVQ addr+0(FP), R8
+	MOVQ oldp+8(FP), AX
+	MOVQ oldu+16(FP), DX
+	MOVQ newp+24(FP), BX
+	MOVQ newu+32(FP), CX
+	LOCK
+	CMPXCHG16B (R8)
```

Comment:
@vmg Do we need to do anything for this to guarantee proper alignment?

Reply:
Yes we do -- we need to align it ourselves or the instruction will trap. :sweat_smile:

Unfortunately there's no way to force compiler alignment at this point, but any alignment issues will be quickly caught in CI as they're thoroughly tested. The good news is that I did not pull this API off my ass: 128 bit atomic have been approved and will be implemented using this exact same API in the next Go release, so we'll be able to switch seamlessly and get automatic compiler alignment for free very soon.